### PR TITLE
Automatically decide when to stop sampling during active learning

### LINKED ISSUE #23292: Automatically decide when to stop sampling during active learning
## Reason
During active learning, right now, if the user request 4000 steps and there are 500 re-training steps, the user effectively gets 3500 samples. @vincentlaboure capability to automatically cut the sampler off when a prescribed size is reached directly from the samplers block will be useful.

## Design
@vincentlaboure has already implemented this for adaptive Monte Carlo samplers. It needs to integrated with the `ActiveLearningMonteCarlo` sampler. With this capability to automatically adjust the total samples on the fly, the user can exactly get 4000 samples.

## Impact
Should not impact existing objects.

### PR DESCRIPTION
**Pull request by vincentlaboure**
closes #23292

## Reason

With active learning, when the Gaussian Process prediction fails, the current time step is 'wasted' and the retraining step doesn't happen until the next time step. 

## Design

Therefore, we keep track of the number of retraining steps to increase the total number of steps taken. A terminator is relied upon to stop the simulation at the right moment. 

## Impact

Enhanced active learning capabilities.

### CODE REVIEW

**Comment by vincentlaboure** [comment]
@somu15 This should be ready for review :)

**Comment by vincentlaboure** [comment]
@somu15 Quick reminder before the PR becomes stale :)

File: modules/stochastic_tools/src/samplers/ActiveLearningMonteCarloSampler.C [hunk 1]
```diff
@@ -55,12 +61,23 @@ ActiveLearningMonteCarloSampler::sampleSetUp(const Sampler::SampleMode /*mode*/)
   if (_check_step == _step)
     return;
 
+  if (_is_sampling_completed)
+    mooseError("Internal bug: the adaptive sampling is supposed to be completed but another sample "
```
Comment (somu15) [inline modules/stochastic_tools/src/samplers/ActiveLearningMonteCarloSampler.C:65]: Please add a test for this for coverage purposes.

File: modules/stochastic_tools/test/tests/reporters/ActiveLearningGP/tests [hunk 2]
```diff
@@ -15,22 +16,32 @@
       min_parallel = 2
       jsondiff = 'multiple_proc_multiple_row_ufunction.json'
       cli_args = "Samplers/mc/num_batch=2 Outputs/file_base='multiple_proc_multiple_row_ufunction'"
-      detail = 'multiple rows in the sampler using the U learning function;'
+      detail = 'multiple rows in the sampler using the U learning function and automatically increase the number of steps when re-training is needed;'
+    []
+    [SingleProc_SingleRow_Ufunction_retraining]
+      type = 'JSONDiff'
```
Comment (somu15) [inline modules/stochastic_tools/test/tests/reporters/ActiveLearningGP/tests:22]: You have increased the `learning_function_threshold` here to better demonstrate the step size increment. Is that right?

**Review by somu15 — CHANGES_REQUESTED** [review]
Looks good! A couple of minor comments. Also, please modify `ActiveLearningGPDecision.md` and `ActiveLearningMonteCarloSampler.md` so that users are aware of this automatic step increments. Thanks for this contribution!

File: modules/stochastic_tools/test/tests/reporters/ActiveLearningGP/tests (hunk 2 above)
Comment (vincentlaboure) [reply modules/stochastic_tools/test/tests/reporters/ActiveLearningGP/tests:22]: It was to make sure re-training was needed

File: modules/stochastic_tools/src/samplers/ActiveLearningMonteCarloSampler.C (hunk 1 above)
Comment (vincentlaboure) [reply modules/stochastic_tools/src/samplers/ActiveLearningMonteCarloSampler.C:65]: What kind of test are you thinking about? Usually, this error is hit when the implementation is incorrect (e.g., when you added a sampler but didn't include it in the list of `adaptiveSamplerNames`). In other words, this is a check to help future developers more than users.

**Review by somu15 — APPROVED** [review]

**Comment by vincentlaboure** [comment]
@zachmprince Any comment on your end?

**Review by zachmprince — APPROVED** [review]

**Comment by lindsayad** [comment]
This pull request is failing the Valgrind Stochastic Tools recipe. See https://civet.inl.gov/job/1470207/

**Pull request merged**